### ISSUE TITLE
"extending a protocol composition is not supported;" when working with shadowed enums

### ISSUE THREAD

**Issue by nmggithub** [swiftlang/swift#76633]
### Description

This is a really weird bug and I think Swift is really getting confused. Please take a look at the reproduction repository, the code itself should (hopefully) be self-explanatory.

### Reproduction

https://github.com/nmggithub/swift-compiler-bug

The first commit is a broken build.
The second commit is a working build.

Simply changing a filename fixes the build. There also doesn't seem to be any sense to what names work and what names don't.

### Expected behavior

Both commits build.

### Environment

```
swift-driver version: 1.115 Apple Swift version 6.0 (swiftlang-6.0.0.9.10 clang-1600.0.26.2)
Target: arm64-apple-macosx15.0
```

**Comment by stackotter** [comment]
I believe that I'm facing the same issue, but with Swift 6.3 and extending a nested class. Exact same warning. I'm seeing the issue in [moreSwift/AndroidKit](https://github.com/moreSwift/AndroidKit). Here's one of the setups causing the issue (simplified from the original code);

### `AppBarLayout.swift

```swift
@JavaClass("com.google.android.material.appbar.AppBarLayout")
open class AppBarLayout: LinearLayout {
    // ...
}
```

### `AppBarLayout+LayoutParams.swift`

```swift
extension AppBarLayout {
    @JavaClass("com.google.android.material.appbar.AppBarLayout$LayoutParams")
    open class LayoutParams: LinearLayout.LayoutParams {
        // ...
    }
}
```

### `AppBarLayout+LayoutParams+ScrollEffect.swift`

```swift
extension AppBarLayout.LayoutParams {
    public typealias ScrollEffect = AppBarLayoutLayoutParamsScrollEffect
}
```

---

I believe that the alphabetic ordering of those file names also lines up with what you were seeing, although I haven't tried reordering the file names yet, because the project does take a while to build.

**Comment by slavapestov** [comment]
It looks like we don't really handle this situation well, where both a superclass and subclass define a nested type with the same name. I agree that the behavior of the compiler should not depend on the order of file names in a module. However, I would also recommend not writing code like this at all, since it is potentially very confusing to the reader.

**Comment by nmggithub** [comment]
> However, I would also recommend not writing code like this at all, since it is potentially very confusing to the reader.

I agree, to some extent. I'll admit, I forget what the larger actual circumstances were that lead to me filing this report. I think the minimal reproduction may, fairly or unfairly, specifically highlight only the bare requirements for the bug (which may look more nonsensical in isolation than in an actual codebase).

(apologies if that comment wasn't directed at me)
